majortom_gateway/gateway_api.py

```diff
@@ -0,0 +1,375 @@
+        message_type = message["type"]
+        logger.debug("From Major Tom: {}".format(message))
+        if message_type == "command":
+            command = Command(message["command"])
+            if self.command_callback is not None:
+                """
+                TODO: Track the task and ensure it completes without errors
+                reference: https://medium.com/@yeraydiazdiaz/asyncio-coroutine-patterns-errors-and-cancellation-3bb422e961ff
+                """
+                asyncio.ensure_future(self.command_callback(command, self))
+            else:
+                asyncio.ensure_future(self.fail_command(
+                    command.id, errors=["No command callback implemented"]))
+        elif message_type == "cancel":
+            if self.cancel_callback is not None:
+                """
+                TODO: Track the task and ensure it completes without errors
+                reference: https://medium.com/@yeraydiazdiaz/asyncio-coroutine-patterns-errors-and-cancellation-3bb422e961ff
+                """
+                asyncio.ensure_future(self.cancel_callback(message["command"]["id"], self))
+            else:
+                asyncio.ensure_future(self.transmit_events(events=[{
+                    "system": None,
+                    "type": "Command Cancellation Failed",
+                    "command_id": message["command"]["id"],
+                    "level": "warning",
+                    "message": "No cancel callback registered. Unable to cancel command."
+                }]))
+        elif message_type == "error":
+            logger.error("Error from Major Tom: {}".format(message["error"]))
+            if self.error_callback is not None:
+                asyncio.ensure_future(self.error_callback(message))
+        elif message_type == "rate_limit":
+            logger.error("Rate limit from Major Tom: {}".format(message["rate_limit"]))
+            if self.rate_limit_callback is not None:
+                asyncio.ensure_future(self.rate_limit_callback(message))
+        elif message_type == "hello":
+            logger.info("Major Tom says hello: {}".format(message))
+        else:
+            logger.warning("Unknown message type {} received from Major Tom: {}".format(
+                message_type, message))
+
+    async def empty_queue(self):
+        while len(self.queued_payloads) > 0 and self.websocket:
+            payload = self.queued_payloads.pop(0)
+            await self.transmit(payload)
+
+    async def transmit(self, payload):
+        if self.websocket:
+            logger.debug("To Major Tom: {}".format(payload))
+            try:
+                await self.websocket.send(json.dumps(payload))
+            except Exception as e:
+                self.websocket = None
+                self.queued_payloads.append(payload)
+        else:
+            # Switch to https://docs.python.org/3/library/asyncio-queue.html
+            self.queued_payloads.append(payload)
```

Comment:
This should check the length of `queued_payloads` and avoid adding if it's over a certain length to avoid filling up RAM. See https://github.com/cantino/major-tom-ruby/blob/master/lib/major_tom/client.rb#L184

Reply:
Good call.

Reply:
Should be here too right?